stream-chat-android-client/src/main/java/io/getstream/chat/android/client/network/NetworkStateProvider.kt

```diff
@@ -0,0 +1,53 @@
+package io.getstream.chat.android.client.network
+
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkRequest
+
+internal class NetworkStateProvider(private val connectivityManager: ConnectivityManager) {
+
+    private val callback = object : ConnectivityManager.NetworkCallback() {
+        override fun onAvailable(network: Network?) {
+            if (!isConnected) {
+                isConnected = true
+                listeners.forEach { it.onConnected() }
+            }
+        }
+
+        override fun onLost(network: Network?) {
+            // Checks whether the network was switched or connection was lost
+            val isConnected = connectivityManager.activeNetworkInfo?.isConnected ?: false
+            if (!isConnected && this@NetworkStateProvider.isConnected) {
+                this@NetworkStateProvider.isConnected = false
+                listeners.forEach { it.onDisconnected() }
+            }
+        }
+    }
+
+    @Volatile
+    private var isConnected: Boolean = false
+    private var listeners = mutableListOf<NetworkStateListener>()
```

Comment:
Seems it requires some synchronization (synchornizedCollections from JavaAPI for example)